hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bloom/HoodieBloomIndex.java

```diff
@@ -119,11 +119,11 @@ private HoodiePairData<HoodieKey, HoodieRecordLocation> lookupIndex(
       HoodiePairData<String, String> partitionRecordKeyPairs, final HoodieEngineContext context,
       final HoodieTable hoodieTable) {
     // Step 1: Obtain records per partition, in the incoming records
-    Map<String, Long> recordsPerPartition = partitionRecordKeyPairs.countByKey();
-    List<String> affectedPartitionPathList = new ArrayList<>(recordsPerPartition.keySet());
+    HoodiePairData<String, Long> recordsPerPartition = partitionRecordKeyPairs.mapToPair(pair -> Pair.of(pair.getLeft(), 1L)).reduceByKey(Long::sum, 1);
```

Comment:
The records per partition is used in 1 of 3 possible flows through the bloom filter and that flow is not enabled by default. My goal here was to make this lazily evaluated and avoid an unnecessary count call in the default case.

Reply:
@vinothchandar this is the main win I see in this PR

Reply:
good find